docs/CompilationExecution.md

```diff
@@ -0,0 +1,115 @@
+ - The `Interpreter` takes the IR and interprets it.
+ Some operations are trivial ---e.g., `1 + 2` or `obj.b` or `arr[0]`--- and do not require any further step to be executed.
+ For data manipulation pipelines, instead, the `Intepreter` needs more.
+ - The `Interpreter` does not know how to execute data pipelines.
+ So, it takes their representation ---i.e., the `Spec`-- and optimizes it by applying subsequent _transformations_, in order to delegates execution to the `Engine`.
+ Those transformations are logical and physical planning.
+ We consider these transformations as further compilation steps, in that they turn a query representation into another.
+ - Finally, the `Engine` executes the optimized representation and outputs the results.
+
+Wrapping up, the actors that play in this process are `Compiler`s ---that turn a representation in some language into another one--- and `Executor`s ---that execute a representation and produce results.
+Both the `Interpreter` and the `Engine` are `Executor`s:
+ - The `Engine` actually executes data manipulation pipelines and produces results;
+ - The `Interpreter` directly executes trivial operations and delegates pipeline execution to the `Engine`.
+ In its essence, it executes a higher level representation for the query than the `Engine`'s.
+ 
+In order to carry out execution, `Executor` need to know some information on a per-query basis (for instance, its memory limits, the request context, etc.).
+This information is embedded into the `ExecutionContext`.
+ 
+These are the interfaces:
+
+```go
+// Go's builtin `context` library.
+import "context"
+
+// Language is a language used to express a Flux query.
+type Language interface {
+    LanguageName() String
+}
+
+// QueryRepresentation contains the actual content of a Flux query expressed in some Language.
+type QueryRepresentation interface {
+    Lang() Language
+}
+
+// Results are the results of a Query.
+type Results interface {}
+// Statistics are the statistics of a Query execution.
+type Statistics interface {}
+
+// Query is an executed Flux query.
+// It provides its representation, results, error, and statistics.
+type Query interface {
+    Results() Results
+    Error() error
+    Stats() Statistics
+}
+
+// The ExecutionContext contains the information necessary for properly executing a query.
+type ExecutionContext interface { 
+    // ... examples of components for the execution context.
+    Context() context.Context
+    MemoryAllocator() Allocator // (Allocator definition is not relevant here)
+    Logger() Logger // (Logger definition is not relevant here)
+}
+
+// Executor executes a query QueryRepresentation and returns a Query given an ExecutionContext.
+type Executor interface {
+    Execute(ExecutionContext, QueryRepresentation) (Query, error)
+    // each Executor can execute query representations in a target Language.
+    ExecutorType() Language
```

Comment:
Just a nitpick but this method doesn't seem like a necessary part of the interface. I would remove it.

Reply:
I don't know if I agree 😅 I think we can decide upon implementation.